app/services/category_hashtag_data_source.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+# Used as a data source via HashtagAutocompleteService to provide category
+# results when looking up a category slug via markdown or searching for
+# categories via the # autocomplete character.
+class CategoryHashtagDataSource
+  def self.icon
+    "folder"
+  end
+
+  def self.category_to_hashtag_item(guardian_categories, category)
+    category = Category.new(category.slice(:id, :slug, :name, :parent_category_id)) if category.is_a?(Hash)
+
+    HashtagAutocompleteService::HashtagItem.new.tap do |item|
+      item.text = category.name
+      item.slug = category.slug
+      item.icon = icon
+      item.relative_url = category.url
+
+      # Single-level category heirarchy should be enough to distinguish between
+      # categories here.
+      item.ref =
+        if category.parent_category_id
+          parent_category =
+            guardian_categories.find { |gc| gc[:id] === category.parent_category_id }
+          category.slug if !parent_category
+
+          parent_slug = parent_category[:slug]
+          "#{parent_slug}:#{category.slug}"
+        else
+          category.slug
+        end
+    end
+  end
+
+  def self.lookup(guardian, slugs)
+    guardian_categories = Site.new(guardian).categories
+    category_slugs_and_ids =
+      slugs.map { |slug| [slug, Category.query_from_hashtag_slug(slug)&.id] }.to_h
+    Category
+      .secured(guardian)
+      .select(:id, :slug, :name, :parent_category_id)
+      .where(id: category_slugs_and_ids.values)
```

Comment:
I dont understand what is the goal of all this manipulation here.
It feels like we could just do:

```
category_slugs_and_ids =  slugs.map { |slug| Category.query_from_hashtag_slug(slug)&.id }
```

Reply:
Yes that sounds better will fix

Reply:
Oh I see why, it's because the old way doesn't accept a Guardian. I will move things around to make this nicer 👍